content/material/2020-01-17-intermediate-dplyr.Rmd

```diff
@@ -0,0 +1,339 @@
+- the argument `values_from`: the values that we will use to fill the rows of the new variable we created
+
+```{r}
+population_wide <- 
+  pivot_wider(population, names_from = country, values_from = population) 
+```
+
+We can run the exact same operation taking advantage of the pipe operator `%>%`
+that we showed in the intro session. If the memory is foggy, imagine this operator as an
+arrow that connects consecutive functions in a smooth way. 
+
+When using the pipe operator we are indicating that the functions will be applied to the 
+dataframe `population`, so we do not need to indicate that into the `pivot_wider()` function arguments.
+
+```{r}
+population_wide <- population %>% 
+  pivot_wider(names_from = country, values_from = population) 
+```
+
+
+#### Exercise 1
+Use the dataframe `us_rent_income` for this exercise. This dataframe shows the estimated value and the 90%
+margin of error for yearly income and monthly rent of each state in the US.
+
+In this case, you want to display the data in wide form to see how estimated income varies across states.
+To do so, each state will become a variable (i.e. column) to which the estimated income will be associated
+
+```{r, eval FALSE}
+us_income_wide <- us_rent_income %>% 
+  #because we are only interested in income and not in rent, we will extract the variable of interest from our database
+  filter(variable == "income") %>% 
+  #we select only the variables of interest to remain in our dataframe
+  select(NAME, variable, estimate) %>% 
+  pivot_wider(names_from = ______, values_from = _____)
+```
+
+### `Pivot_longer()`
+
+We can also do the opposite: a shift from a dataframe in wide format to a dataframe in long format. 
+Let's consider the `population_wide` dataframe that we just created as the starting point. 
+Let's say that we want to reshape this data to a format in which all countries are listed under a single variable, e.g. `country_name`, and the population data are listed under a variable called `population_data`. For this operation, we need to use the function `pivot_longer()`
+
+In this case, we want to specify:
+- the argument `names_to`: the name of the new variable under which we want to group our columns, 
+- the argument `values_to`: the name of the new variable under which we want to group the values of our variables.
+
+We run this function...
+
+```{r}
+population_wide %>%
+  pivot_longer(names_to = "Country", values_to = "Population")
+```
+
+... but we get an error message!
+
+This is because we did not specify which columns to shift into a long format.
+
+  > In `pivot_wider()`, if no argument is given to `id_cols` it defaults to select all columns. However, `pivot_longer` has no default for `cols`, and will throw an error if no argument is given.
+
+We could select every column, with the tidyselect function `everything()`. This is not what want (..., but you should try it to see why!). 
+
```

Comment:
Hi Johannes, great idea! 
Maybe we can add a code chunk were we give and example on how to use the function everything().
I think it would be helpful

Reply:
Great, i added it in 117a1de